File: rust-martian-ddd/martian-robot-ddd/src/application/in_memory_scents.rs
use crate::domain::{orientation::Orientation, position::Position, scents::ScentRepository};
use std::collections::HashSet;

#[derive(Default)]
pub struct InMemoryScents {
    set: HashSet<(i32, i32, Orientation)>,
}

impl ScentRepository for InMemoryScents {
    fn has_scent(&self, pos: Position, o: Orientation) -> bool {
        self.set.contains(&(pos.x, pos.y, o))
    }
    fn add_scent(&mut self, pos: Position, o: Orientation) {
        self.set.insert((pos.x, pos.y, o));
    }
}

File: java-martian-ddd/src/InMemoryScents.java
import java.util.*;

final class InMemoryScents implements ScentRepository {
    private final Set<String> set = new HashSet<>();

    public boolean has(Position p, Orientation o) {
        return set.contains(p.x + "," + p.y + "," + o);
    }

    public void add(Position p, Orientation o) {
        set.add(p.x + "," + p.y + "," + o);
    }
}
